src/SAWScript/Crucible/Common/MethodSpec.hs

```diff
@@ -71,121 +155,78 @@ stateCond :: PrePost -> String
-data SetupValue ext where
-  SetupVar    :: AllocIndex -> SetupValue ext
-  SetupTerm   :: TypedTerm -> SetupValue ext
-  SetupNull   :: B (HasSetupNull ext) -> SetupValue ext
-  -- | If the 'Bool' is 'True', it's a (LLVM) packed struct
-  SetupStruct :: B (HasSetupStruct ext) -> Bool -> [SetupValue ext] -> SetupValue ext
-  SetupArray  :: B (HasSetupArray ext) -> [SetupValue ext] -> SetupValue ext
-  SetupElem   :: B (HasSetupElem ext) -> SetupValue ext -> Int -> SetupValue ext
-  SetupField  :: B (HasSetupField ext) -> SetupValue ext -> String -> SetupValue ext
-  SetupCast   :: B (HasSetupCast ext) -> SetupValue ext -> CastType ext -> SetupValue ext
-  SetupUnion  :: B (HasSetupUnion ext) -> SetupValue ext -> String -> SetupValue ext
-
-  -- | A pointer to a global variable
-  SetupGlobal :: B (HasSetupGlobal ext) -> String -> SetupValue ext
-  -- | This represents the value of a global's initializer.
-  SetupGlobalInitializer ::
-    B (HasSetupGlobalInitializer ext) -> String -> SetupValue ext
-
--- | This constraint can be solved for any ext so long as '()' and 'Void' have
---   the constraint. Unfortunately, GHC can't (yet?) reason over the equations
---   in our closed type family, and realize that
-type SetupValueHas (c :: Type -> Constraint) ext =
-  ( c (B (HasSetupNull ext))
-  , c (B (HasSetupStruct ext))
-  , c (B (HasSetupArray ext))
-  , c (B (HasSetupElem ext))
-  , c (B (HasSetupField ext))
-  , c (B (HasSetupCast ext))
-  , c (B (HasSetupUnion ext))
-  , c (B (HasSetupGlobal ext))
-  , c (B (HasSetupGlobalInitializer ext))
-  , c (CastType ext)
-  )
-
-deriving instance (SetupValueHas Show ext) => Show (SetupValue ext)
-
--- TypedTerm is neither Eq nor Ord
--- deriving instance (SetupValueHas Eq ext) => Eq (SetupValue ext)
--- deriving instance (SetupValueHas Ord ext) => Ord (SetupValue ext)
+-- | A singleton type that encodes the extensions used for each SAW backend.
+-- Matching on a value of type @'SAWExt' ext@ allows one to learn which backend
+-- is currently in use.
+data SAWExt :: Type -> Type where
+  LLVMExt :: SAWExt (LLVM arch)
+  JVMExt  :: SAWExt JVM
+  MIRExt  :: SAWExt MIR
+
+-- | This allows checking which SAW backend we are using at runtime.
+class IsExt ext where
+  sawExt :: SAWExt ext
+
+instance IsExt (LLVM arch) where
+  sawExt = LLVMExt
+instance IsExt JVM where
+  sawExt = JVMExt
+instance IsExt MIR where
+  sawExt = MIRExt
 
 -- | Note that most 'SetupValue' concepts (like allocation indices)
 --   are implementation details and won't be familiar to users.
---   Consider using 'resolveSetupValue' and printing an 'LLVMVal'
---   with @PP.pretty@ instead.
-ppSetupValue :: Show (CastType ext) => SetupValue ext -> PP.Doc ann
+--   Consider using 'resolveSetupValue' and printing the language-specific value
+--   (e.g., an 'LLVMVal') with @PP.pretty@ instead.
+ppSetupValue :: forall ext ann. IsExt ext => SetupValue ext -> PP.Doc ann
```

Comment:
Adding this `IsExt` constraint to `ppSetupValue` is probably the largest change in this patch, aside from changing the TTG-related type families, as I have to perform case analysis on which extension is being used in order to pretty-print `SetupValue`s properly. This is primarily what motivated moving code around so much in this patch, as I needed to be able to import all of the language backends from `SAWScript.Crucible.Common.MethodSpec` without creating import cycles.